### PR TITLE
Fix the duplicate glyphs in Inheritance margin

### PR DESCRIPTION
**Pull request by Cosifne**
Fix: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1988154/
Currently, you might find there are duplicate glyphs in inheritance margin like:
image
(There are three controls created in the canvas, all of them are same)

Reason:
Our tagger infra has switched to a three-tagger stucture.
Tagger 1 -> Tag the [visible region + 10 line, visible region - 10 line]
Tagger 2 -> Tag the [visible region + 10 line, visible region + 100 line]
Tagger 3 -> Tag the [visible region - 100 line, visible region - 10 line]

This change causes problem [here](https://github.com/dotnet/roslyn/blob/0447e0c1549f0f9a2706853172f00507b352773f/src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs#L82)

So computing the symbol to tag, inheritance margin service would call `root.DescendantNodes(spanToSearch)` to find all the nodes **intersect** with `spanToSearch`.
So suppose you have a class, and the visible region is around `class A`
```
//1k lines comments
class A
{
// 1k lines body
}
```
Tagger 1 would tag class `A`. (Because the `spanToSearch` is the visible span)
Tagger 2 would also tag class `A`, because the top lines span also intersects with `class A`
Tagger 3 would also tag class `A`, because the empty body span also intersects with `class A`

The solution is to only create the tag when `spanToSearch` contains the member's identifier.

### CODE REVIEW

File: src/EditorFeatures/Test/InheritanceMargin/InheritanceMarginTests.cs [hunk 1]
```diff
@@ -523,9 +524,9 @@ public Task TestCSharpSpecialMember(string memberDeclaration, TestHost testHost)
         var markup = $@"
         public abstract class {{|target1:Bar1|}}
         {{}}
-        public class Bar : Bar1
+        public class {{|{SearchAreaTag}:Bar : Bar1
```
Comment (Cosifne) [inline src/EditorFeatures/Test/InheritanceMargin/InheritanceMarginTests.cs:527]: This test is modified a little bit.
The prev test is asserting that no inheritanceMargin tag should be created for special C# methods. (e.g. constructor/deconsctructor/operator)

So the assertion is just assert one tag is created for the type (`class Bar1` here)

It only chooses to search the `memberDeclaration` region so only tags for `Bar1` would be created.

However, after the change, we would only return the tag within the search span. So expand the `SearchAreaTag` area here to include `class Bar1`
No real change here.

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs [hunk 2]
```diff
@@ -96,6 +96,15 @@ internal abstract partial class AbstractInheritanceMarginService
                 if (member == null || !CanHaveInheritanceTarget(member))
                     continue;
 
+                // Make sure the identifier declaration's span is within the spanToSearch.
+                // This is important because for example, for a class with big body, when the tagger is asking for the tag within the body of the class,
+                // we don't want to return the tag of the class identifier.
+                var declarationToken = GetDeclarationToken(memberDeclarationNode);
+                if (!spanToSearch.Contains(declarationToken.Span))
```
Comment (Cosifne) [inline src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:103]: `spanToSearch` comes from https://github.com/dotnet/roslyn/blob/0447e0c1549f0f9a2706853172f00507b352773f/src/VisualStudio/Core/Def/InheritanceMargin/InheritanceMarginTaggerProvider.cs#L91C28-L91C37

So it is assuming the tagger infra to always ask the correct region to tag.
(Which I believe it should)

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InheritanceMargin/AbstractInheritanceMarginService_Helpers.cs:103]: can you move this above the semantic checks above.  this syntax check is much faster than the symbol checks.

**Comment by DoctorKrolic** [comment]
Fixes https://github.com/dotnet/roslyn/issues/73615?

**Comment by Cosifne** [comment]
> Fixes #73615?

Yes it should fix the problem

**Pull request merged**